normalization/normalization.md

```diff
@@ -0,0 +1,77 @@
+
+A way to make this is 3NF compliant is to divide this table further. A grades table with the columns `entry_id` and `score`, and a keys table with `entry_id`, `student_id`, and `professor_id`.
+
+4) 
+
+There are a couple of difficulties linked with normalization. It can be difficult to design, that is, it requires some detail and thought upfront before designing the database. This also may introduce some issues if the database is to changed or scale in the future, introducing some complexity in the planning stage. A mistake in design, or an unforseen issue during design could bring confusing and sub-optimal consecuences.
+
+Failing to determine or introducing a mistake, for example, in the relationships between keys and the attributes during the design phase can be disastrous, especially after data has been saved into the database itself. Data that has been committed might have to be shifted around, the creation or deletion of columns opens the process up to mistakes and data loss. 
+
+Another issue is that, because we are breaking up tables in order to avoid redundancy, there could be a table with very little data due to normalization, yet necessary in order to keep with whatever level of Normal Form we want to comply with. Because of this, there will be a need to use joins to get a small amount of data, which can increase the complexity of the queries themselves.
+
+Because you are spreading out information to normalize a database, the performance can be reduced due to the increased amount of joins.
+
+5)
+
+So, for example, in the previous table of grades, professors, students, and keys that were normalized, a strategy is to perhaps, loosen the level of Normalization.
+
+I believe that 1NF data flattening is a good practice, as well as applying 2NF in order to avoid duplicating information, perhaps 3NF might be a bit strict, as the grades table already has a direct, unique entry for all key candidates in the table.
+
+That is, in this stage of normalization where the data is 2NF compliant, the 2NF compliant grades table contains `entry_id`, `student_id`, `professor_id`, and `score` columns. Every `entry_id` contains a unique `{student_id, professor_id}` value pair, as well as unique `score` value. Yes, there is a transitive relationship too between `entry_id` and `score`. However, it's one less join we'd have to make if we only wanted to get the average of a student if only we just leave it as 2NF compliant. Ultimately, all information in each table, is unique, or in combination, is unique to the `entry_id`, and hence 3NF could be removed and only relax normalization to 2NF.
+
+This would be a good strategy in my limited opinion.
+
+6) We must discuss!!!!!
```

Comment:
shouldn't have this commit in this submission ;)

Reply:
ohh! :man_facepalming: